### PR TITLE
V4 self service contract

### PR DESCRIPTION
**Pull request by jamesmorgan**
### New Changes

* [x] Extract common interfaces
* [x] Allow settings start and end date
* [x] Allow setting split commission
* [x] Fix existing test runner
* [x] Tests passing
* [x] deploy to ropsten
* [x] deploy to rinkeby
* [ ] Update KODA tools
* [ ] deploy to mainnet
* [ ] update API

### CODE REVIEW

File: contracts/v2/self-service/SelfServiceEditionCurationV4.sol [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+  whenNotPaused
+  returns (uint256 _editionNumber)
+  {
+    require(frequencyControls.canCreateNewEdition(msg.sender), 'Sender currently frozen out of creation');
+    require(_artistCommission.add(_optionalSplitRate).add(koCommission) <= 100, "Total commission exceeds 100");
+
+    uint256 editionNumber = _createEdition(msg.sender, _enableAuction, _totalAvailable, _priceInWei, _artistCommission, _tokenUri);
+
+    if (_startDate > 0) {
+      kodaV2.updateStartDate(editionNumber, _startDate);
```
Comment (andygray) [inline contracts/v2/self-service/SelfServiceEditionCurationV4.sol]: can start date be in the past?

File: contracts/v2/self-service/SelfServiceEditionCurationV4.sol [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    accessControls = _accessControls;
+    frequencyControls = _frequencyControls;
+  }
+
+  /**
+   * @dev Called by artists, create new edition on the KODA platform
+   */
+  function createEdition(
+    bool _enableAuction,
+    address _optionalSplitAddress,
```
Comment (andygray) [inline contracts/v2/self-service/SelfServiceEditionCurationV4.sol]: Let's face it - they are going to want to create multi-splits between x artists. Worth addressing that now?

File: contracts/v2/self-service/SelfServiceEditionCurationV4.sol [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    require(_totalAvailable > 0, "Must be at least one available in edition");
+    require(_totalAvailable <= maxEditionSize, "Must not exceed max edition size");
+
+    // Enforce min price
+    require(_priceInWei >= minPricePerEdition, "Price must be greater than minimum");
+
+    // If we are the owner, skip this artists check
+    if (msg.sender != owner) {
+      // Enforce who can call this
+      require(accessControls.isEnabledForAccount(_artist), "Only allowed artists can create editions for now");
```
Comment (andygray) [inline contracts/v2/self-service/SelfServiceEditionCurationV4.sol]: add the if into the require maybe?

File: contracts/v2/self-service/SelfServiceEditionCurationV4.sol [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+    uint256 _priceInWei,
+    uint256 _artistCommission,
+    string _tokenUri
+  )
+  internal
+  returns (bool) {
+    return kodaV2.createActiveEdition(
+      _editionNumber,
+      0x0, // _editionData - no edition data
+      1, // _editionType - KODA always type 1
```
Comment (andygray) [inline contracts/v2/self-service/SelfServiceEditionCurationV4.sol]: feels like we will start using `type` more. Pass in?

Comment (jamesmorgan) [reply contracts/v2/self-service/SelfServiceEditionCurationV4.sol]: Good idea

File: contracts/v2/self-service/SelfServiceEditionCurationV4.sol (hunk 2 above)
Comment (jamesmorgan) [reply contracts/v2/self-service/SelfServiceEditionCurationV4.sol]: Agree, not sure how we do that easily though? Ideas?

File: contracts/v2/self-service/SelfServiceEditionCurationV4.sol (hunk 1 above)
Comment (jamesmorgan) [reply contracts/v2/self-service/SelfServiceEditionCurationV4.sol]: No benefit to it being in the past I think

**Pull request merged**